packages/dds/merge-tree/src/client.ts

```diff
@@ -1021,17 +1031,49 @@ export class Client {
     }
 
     getContainingSegment<T extends ISegment>(pos: number, op?: ISequencedDocumentMessage) {
-        let seq: number;
-        let clientId: number;
-        if (op) {
-            clientId = this.getOrAddShortClientId(op.clientId);
-            seq = op.referenceSequenceNumber;
-        } else {
-            const segWindow = this.mergeTree.getCollabWindow();
-            seq = segWindow.currentSeq;
-            clientId = segWindow.clientId;
+        const args = this.getClientSequenceArgsForMessage(op);
+        return this.mergeTree.getContainingSegment<T>(pos, args.referenceSequenceNumber, args.clientId);
+    }
+
+    /**
+     * Returns the segment and offset for creating a SlideOnRemove reference position
+     * in response to a remote op.
+     * Will return a location which is slid in an eventually consistent way.
+     * @param pos - The remote position
+     * @param op - The remote op
+     * @returns - segment and offset at which to create a SlideOnRemove reference position.
+     */
+    getSlideOnRemoveReferencePosition(pos: number, op: ISequencedDocumentMessage) {
+        const args = this.getClientSequenceArgsForMessage(op);
+        const segoff = this.mergeTree.getSlideOnRemoveReferenceSegmentAndOffset(
+            pos, args.referenceSequenceNumber, args.clientId);
+        if (!segoff.segment || segoff.offset === undefined || segoff.offset < 0) {
```

Comment:
This validation is overly strict: there are cases where I think we must handle no segment. I'd propose this check instead:

```suggestion
        if (segoff.offset === undefined || segoff.offset < 0 || (!segoff.segment && segoff.offset > 0)) {
```

Here's an example interval collection test case that fails:

```
        it("tolerates creation of an interval with no segment due to concurrent delete", () => {
            sharedString.insertText(0, "ABCDEF");
            const collection1 = sharedString.getIntervalCollection("test");
            const collection2 = sharedString2.getIntervalCollection("test");
            containerRuntimeFactory.processAllMessages();
            sharedString2.removeRange(0, sharedString2.getLength());
            collection1.add(1, 1, IntervalType.SlideOnRemove);
            containerRuntimeFactory.processAllMessages();
            assertIntervals(sharedString, collection1, [{ start: -1, end: -1 }]);
            assertIntervals(sharedString2, collection2, [{ start: -1, end: -1 }]);
        });
```

Reply:
To actually get that test case passing requires easing another assert in interval collection.

I'm also OK addressing this in a follow-up.. just thought pointing it out here would be worthwhile in case it means we want to reconsider our policy on detaching refs when there's no segment to slide to.